pkg/store/catalog.go

```diff
@@ -124,3 +148,109 @@ func (c *Catalog) GetPlanOf(serviceInstanceSpec *sc_v1b1.ServiceInstanceSpec) (*
 	}
 }
+
+func makePlanSchemaKey(plan *sc_v1b1.ClusterServicePlan, action planSchemaAction) planSchemaKey {
+	return planSchemaKey(fmt.Sprintf("%s/%s/%s", plan.Name, action))
+}
+
+func (c *Catalog) getSchemaCache(plan *sc_v1b1.ClusterServicePlan, action planSchemaAction) (*gojsonschema.Schema, bool) {
+	key := makePlanSchemaKey(plan, action)
+
+	c.schemasRWMutex.RLock()
+	defer c.schemasRWMutex.RUnlock()
+	if schemaWithRv, ok := c.schemas[key]; ok && schemaWithRv.resourceVersion == plan.ResourceVersion {
+		return schemaWithRv.schema, true
+	} else {
+		// nil is a valid entry in the cache
+		return nil, false
+	}
+}
+
+func (c *Catalog) setSchemaCache(plan *sc_v1b1.ClusterServicePlan, action planSchemaAction, schema *gojsonschema.Schema) {
+	key := makePlanSchemaKey(plan, action)
+
+	c.schemasRWMutex.Lock()
+	defer c.schemasRWMutex.Unlock()
+	c.schemas[key] = schemaWithResourceVersion{plan.ResourceVersion, schema}
+}
+
+func (c *Catalog) getParsedSchema(plan *sc_v1b1.ClusterServicePlan, action planSchemaAction) (*gojsonschema.Schema, error) {
+	if schema, ok := c.getSchemaCache(plan, action); ok {
+		return schema, nil
+	}
+
+	var rawSchema *runtime.RawExtension
+	switch action {
+	case instanceCreateAction:
+		rawSchema = plan.Spec.ServiceInstanceCreateParameterSchema
+	case instanceUpdateAction:
+		rawSchema = plan.Spec.ServiceInstanceUpdateParameterSchema
+	case bindingCreateAction:
+		rawSchema = plan.Spec.ServiceBindingCreateParameterSchema
+	default:
+		return nil, errors.Errorf("plan action %q not understood", action)
+	}
+
+	var schema *gojsonschema.Schema
+	if rawSchema == nil {
+		schema = nil
+	} else {
+		var err error
+		schema, err = gojsonschema.NewSchema(gojsonschema.NewBytesLoader(rawSchema.Raw))
+		if err != nil {
+			return nil, errors.Wrapf(err,
+				"cannot parse json schema for plan %q on broker %q",
+				plan.Spec.ExternalName, plan.Spec.ClusterServiceBrokerName)
+		}
+	}
+
+	c.setSchemaCache(plan, action, schema)
+	return schema, nil
+}
+
+func (c *Catalog) ValidateServiceInstanceSpec(serviceInstanceSpec *sc_v1b1.ServiceInstanceSpec) error {
+	if len(serviceInstanceSpec.ParametersFrom) > 0 {
+		return errors.New("cannot validate ServiceInstanceSpec which has a ParametersFrom block (insufficient information)")
+	}
+
+	servicePlan, err := c.GetPlanOf(serviceInstanceSpec)
+	if err != nil {
+		return err
+	}
+
+	// We ignore the update schema here and assume it's equivalent to
+	// create (since kubernetes/service catalog can't properly distinguish
+	// them anyway as there are currently no true PATCH updates).
+	schema, err := c.getParsedSchema(servicePlan, instanceCreateAction)
+	if err != nil {
+		return err
+	}
+	if schema == nil {
+		// no schema to validate against anyway
+		return nil
+	}
+	var parameters []byte
+	if serviceInstanceSpec.Parameters != nil {
+		parameters = serviceInstanceSpec.Parameters.Raw
+	} else {
+		parameters = []byte{}
```

Comment:
Not needed? `nil` slice is a valid slice.

Reply:
schema validator says it doesn't like it. I've made a small change.